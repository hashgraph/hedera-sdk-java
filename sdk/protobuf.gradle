dependencies {
    implementation 'io.grpc:grpc-protobuf:1.50.2'

    // Protobuf Lite is used to maintain easy compatibility with Android
    // https://github.com/protocolbuffers/protobuf/blob/master/java/lite.md
    api "com.google.protobuf:protobuf-java:3.21.8"
}

protobuf {
    generateProtoTasks {
        all()*.plugins {
            grpc {}
        }
    }
}
