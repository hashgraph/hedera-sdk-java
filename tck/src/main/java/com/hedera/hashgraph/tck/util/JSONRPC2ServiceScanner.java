package com.hedera.hashgraph.tck.util;

import com.hedera.hashgraph.tck.annotation.JSONRPC2Service;
import com.thetransactioncompany.jsonrpc2.server.*;
import org.springframework.context.ApplicationContext;

/**
 * Utility class to register all {@link JSONRPC2Service} annotated classes
 * to the JSON-RPC dispatcher
 */
public class JSONRPC2ServiceScanner {
    private JSONRPC2ServiceScanner() {
        // private constructor for utility class
    }

    public static void registerServices(Dispatcher dispatcher, ApplicationContext context) {
        String[] serviceNames = context.getBeanNamesForAnnotation(JSONRPC2Service.class);
        if (serviceNames != null) { // NOSONAR
            for (String serviceName : serviceNames) {
                Object service = context.getBean(serviceName);
                dispatcher.register((RequestHandler) service);
            }
        }
    }
}
