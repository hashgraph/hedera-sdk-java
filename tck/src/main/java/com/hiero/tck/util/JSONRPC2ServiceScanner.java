/*-
 *
 * Hedera Java SDK
 *
 * Copyright (C) 2024 Hedera Hashgraph, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
package com.hiero.tck.util;

import com.hiero.tck.annotation.JSONRPC2Service;
import com.thetransactioncompany.jsonrpc2.server.*;
import org.springframework.context.ApplicationContext;

/**
 * Utility class to register all {@link JSONRPC2Service} annotated classes
 * to the JSON-RPC dispatcher
 */
public class JSONRPC2ServiceScanner {
    private JSONRPC2ServiceScanner() {
        // private constructor for utility class
    }

    public static void registerServices(Dispatcher dispatcher, ApplicationContext context) {
        String[] serviceNames = context.getBeanNamesForAnnotation(JSONRPC2Service.class);
        if (serviceNames != null) { // NOSONAR
            for (String serviceName : serviceNames) {
                Object service = context.getBean(serviceName);
                dispatcher.register((RequestHandler) service);
            }
        }
    }
}
