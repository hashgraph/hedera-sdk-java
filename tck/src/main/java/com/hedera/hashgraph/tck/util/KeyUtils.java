package com.hedera.hashgraph.tck.util;

import com.hedera.hashgraph.sdk.Key;
import com.hedera.hashgraph.sdk.PrivateKey;
import com.hedera.hashgraph.sdk.PublicKey;

public final class KeyUtils {

    // TODO: cover with test(s)
    public static Key getKeyFromStringDER(String keyStringDER) {
        try {
            return PublicKey.fromStringDER(keyStringDER);
        } catch (Exception e) {
            return PrivateKey.fromStringDER(keyStringDER);
        }
    }
}
