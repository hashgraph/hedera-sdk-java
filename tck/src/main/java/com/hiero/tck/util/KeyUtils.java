/*-
 *
 * Hedera Java SDK
 *
 * Copyright (C) 2024 Hedera Hashgraph, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
package com.hiero.tck.util;

import com.google.protobuf.ByteString;
import com.google.protobuf.InvalidProtocolBufferException;
import com.hiero.sdk.Key;
import com.hiero.sdk.PrivateKey;
import com.hiero.sdk.PublicKey;

public final class KeyUtils {

    public enum KeyType {
        ED25519_PRIVATE_KEY("ed25519PrivateKey"),
        ED25519_PUBLIC_KEY("ed25519PublicKey"),
        ECDSA_SECP256K1_PRIVATE_KEY("ecdsaSecp256k1PrivateKey"),
        ECDSA_SECP256K1_PUBLIC_KEY("ecdsaSecp256k1PublicKey"),
        LIST_KEY("keyList"),
        THRESHOLD_KEY("thresholdKey"),
        EVM_ADDRESS_KEY("evmAddress");

        private final String keyString;

        KeyType(String keyString) {
            this.keyString = keyString;
        }

        public static KeyType fromString(String keyString) {
            for (KeyType type : KeyType.values()) {
                if (type.keyString.equals(keyString)) {
                    return type;
                }
            }
            throw new IllegalArgumentException("Unknown key type: " + keyString);
        }
    }

    public static Key getKeyFromString(String keyString) throws InvalidProtocolBufferException {
        try {
            return PublicKey.fromStringDER(keyString);
        } catch (Exception e) {
            try {
                return PrivateKey.fromStringDER(keyString);
            } catch (Exception ex) {
                return Key.fromBytes(ByteString.fromHex(keyString).toByteArray());
            }
        }
    }
}
