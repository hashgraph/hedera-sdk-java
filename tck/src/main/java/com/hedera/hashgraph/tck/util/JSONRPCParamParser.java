// SPDX-License-Identifier: Apache-2.0
package com.hedera.hashgraph.tck.util;

import com.hedera.hashgraph.tck.methods.sdk.param.CommonTransactionParams;
import com.hedera.hashgraph.tck.methods.sdk.param.CustomFee;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import net.minidev.json.JSONArray;
import net.minidev.json.JSONObject;

public class JSONRPCParamParser {

    public static Optional<CommonTransactionParams> parseCommonTransactionParams(Map<String, Object> jrpcParams)
            throws Exception {
        return parseJsonObject(jrpcParams, "commonTransactionParams", CommonTransactionParams::parse);
    }

    public static Optional<List<CustomFee>> parseCustomFees(Map<String, Object> jrpcParams) throws Exception {
        return parseJsonArray(jrpcParams, "customFees", jsonObj -> new CustomFee().parse(jsonObj));
    }

    private static <T> Optional<T> parseJsonObject(
            Map<String, Object> params, String key, ThrowingFunction<JSONObject, T> parser) throws Exception {
        if (!params.containsKey(key)) {
            return Optional.empty();
        }
        JSONObject jsonObject = (JSONObject) params.get(key);
        return Optional.of(parser.apply(jsonObject));
    }

    private static <T> Optional<List<T>> parseJsonArray(
            Map<String, Object> params, String key, ThrowingFunction<JSONObject, T> elementParser) throws Exception {
        if (!params.containsKey(key)) {
            return Optional.empty();
        }
        JSONArray jsonArray = (JSONArray) params.get(key);
        List<T> results = new ArrayList<>();

        for (Object o : jsonArray) {
            JSONObject jsonObject = (JSONObject) o;
            results.add(elementParser.apply(jsonObject));
        }
        return Optional.of(results);
    }

    @FunctionalInterface
    private interface ThrowingFunction<T, R> {
        R apply(T t) throws Exception;
    }
}
