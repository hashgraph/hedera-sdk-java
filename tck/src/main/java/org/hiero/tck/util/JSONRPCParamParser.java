// SPDX-License-Identifier: Apache-2.0
package org.hiero.tck.util;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import net.minidev.json.JSONArray;
import net.minidev.json.JSONObject;
import org.hiero.tck.exception.JSONRPCParseException;
import org.hiero.tck.methods.sdk.param.CommonTransactionParams;
import org.hiero.tck.methods.sdk.param.CustomFee;

public class JSONRPCParamParser {

    public static Optional<CommonTransactionParams> parseCommonTransactionParams(Map<String, Object> jrpcParams) {
        try {
            return parseJsonObject(jrpcParams, "commonTransactionParams", CommonTransactionParams::parse);
        } catch (Exception e) {
            throw new JSONRPCParseException("Failed to parse CommonTransactionParams", e);
        }
    }

    public static Optional<List<CustomFee>> parseCustomFees(Map<String, Object> jrpcParams) {
        try {
            return parseJsonArray(jrpcParams, "customFees", jsonObj -> {
                try {
                    return new CustomFee().parse(jsonObj);
                } catch (Exception e) {
                    throw new JSONRPCParseException("Failed to parse CustomFee", e);
                }
            });
        } catch (Exception e) {
            throw new JSONRPCParseException("Failed to parse CustomFees array", e);
        }
    }

    private static <T> Optional<T> parseJsonObject(
            Map<String, Object> params, String key, ThrowingFunction<JSONObject, T> parser) throws Exception {
        if (!params.containsKey(key)) {
            return Optional.empty();
        }
        JSONObject jsonObject = (JSONObject) params.get(key);
        return Optional.of(parser.apply(jsonObject));
    }

    private static <T> Optional<List<T>> parseJsonArray(
            Map<String, Object> params, String key, ThrowingFunction<JSONObject, T> elementParser) throws Exception {
        if (!params.containsKey(key)) {
            return Optional.empty();
        }
        JSONArray jsonArray = (JSONArray) params.get(key);
        List<T> results = new ArrayList<>();

        for (Object o : jsonArray) {
            JSONObject jsonObject = (JSONObject) o;
            results.add(elementParser.apply(jsonObject));
        }
        return Optional.of(results);
    }

    @FunctionalInterface
    private interface ThrowingFunction<T, R> {
        R apply(T t) throws Exception;
    }
}
