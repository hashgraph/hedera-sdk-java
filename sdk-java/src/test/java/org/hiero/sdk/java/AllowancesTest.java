// SPDX-License-Identifier: Apache-2.0
package org.hiero.sdk.java;

import static org.assertj.core.api.Assertions.assertThat;

import com.google.protobuf.InvalidProtocolBufferException;
import io.github.jsonSnapshot.SnapshotMatcher;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

public class AllowancesTest {
    @BeforeAll
    public static void beforeAll() {
        SnapshotMatcher.start(Snapshot::asJsonString);
    }

    @AfterAll
    public static void afterAll() {
        SnapshotMatcher.validateSnapshots();
    }

    TokenAllowance spawnTokenAllowance() {
        return new TokenAllowance(
                TokenId.fromString("1.2.3"), AccountId.fromString("4.5.6"), AccountId.fromString("5.5.5"), 777);
    }

    TokenNftAllowance spawnNftAllowance() {
        List<Long> serials = new ArrayList<>();
        serials.add(123L);
        serials.add(456L);
        return new TokenNftAllowance(
                TokenId.fromString("1.1.1"),
                AccountId.fromString("2.2.2"),
                AccountId.fromString("3.3.3"),
                null,
                serials,
                null);
    }

    TokenNftAllowance spawnAllNftAllowance() {
        return new TokenNftAllowance(
                TokenId.fromString("1.1.1"),
                AccountId.fromString("2.2.2"),
                AccountId.fromString("3.3.3"),
                null,
                Collections.emptyList(),
                true);
    }

    HbarAllowance spawnHbarAllowance() {
        return new HbarAllowance(AccountId.fromString("1.1.1"), AccountId.fromString("2.2.2"), new Hbar(3));
    }

    @Test
    void shouldSerialize() {
        SnapshotMatcher.expect(
                        spawnHbarAllowance().toString(),
                        spawnTokenAllowance().toString(),
                        spawnNftAllowance().toString(),
                        spawnAllNftAllowance().toString())
                .toMatchSnapshot();
    }

    @Test
    void shouldBytes() throws InvalidProtocolBufferException {
        var hbar1 = spawnHbarAllowance();
        var token1 = spawnTokenAllowance();
        var nft1 = spawnNftAllowance();
        var allNft1 = spawnAllNftAllowance();
        var hbar2 = HbarAllowance.fromBytes(hbar1.toBytes());
        var token2 = TokenAllowance.fromBytes(token1.toBytes());
        var nft2 = TokenNftAllowance.fromBytes(nft1.toBytes());
        var allNft2 = TokenNftAllowance.fromBytes(allNft1.toBytes());
        assertThat(hbar2.toString()).isEqualTo(hbar1.toString());
        assertThat(token2.toString()).isEqualTo(token1.toString());
        assertThat(nft2.toString()).isEqualTo(nft1.toString());
        assertThat(allNft2.toString()).isEqualTo(allNft1.toString());
    }
}
