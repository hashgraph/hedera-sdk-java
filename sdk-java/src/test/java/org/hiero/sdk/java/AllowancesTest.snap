org.hiero.sdk.java.AllowancesTest.shouldSerialize=[
  "HbarAllowance{ownerAccountId=1.1.1, spenderAccountId=2.2.2, amount=3 ‚Ñè}",
  "TokenAllowance{tokenId=1.2.3, ownerAccountId=4.5.6, spenderAccountId=5.5.5, amount=777}",
  "TokenNftAllowance{tokenId=1.1.1, ownerAccountId=2.2.2, spenderAccountId=3.3.3, delegatingSpender=null, serials=[123, 456]}",
  "TokenNftAllowance{tokenId=1.1.1, ownerAccountId=2.2.2, spenderAccountId=3.3.3, delegatingSpender=null, allSerials=true}"
]
